GSoC Objective-1 Rewrite the TARDIS visualisation module using Panel

rudrakshnalbalwar opened the pull request:
Type: 🚀 Feature | ☣️ Breaking Change | 📝 Documentation

🔹 This PR migrates the TARDIS visualization module to Panel, eliminating dependencies on qgrid, qgridnext, and IPython. This transition improves compatibility with Sphinx documentation while maintaining full interactivity.

🔹 The hierarchical dataset is now visualized as interlinked tables using SimulationShellInfo:

Selecting a row in the shell data table dynamically updates the element abundances table.
Further selection in the element abundances table updates additional tables showing ion count and level count.
🔹 The widget is fully operational in:

Jupyter Notebooks (inline execution)

### :pushpin: Resources

Full Video [watch here]
https://drive.google.com/drive/folders/1YX5T1tSHyy8AaJRWUM8WE__Y4BRUfObF?usp=sharing

### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)


### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

atharva-2001 commented on tardis/visualization/widgets/shell_info.py:
```diff
@@ -224,302 +191,201 @@ def __init__(self, hdf_fpath):
 def shell_info_from_hdf(hdf_fpath):
-    """Create shell info widget from a simulation HDF file
-
-    Parameters
-    ----------
-    hdf_fpath : str
-        A valid path to a simulation HDF file (HDF file must be created
-        from a TARDIS Simulation object using :code:`to_hdf` method with
-        default arguments)
-
-    Returns
```
Why remove them?

rudrakshnalbalwar commented on tardis/visualization/widgets/shell_info.py:
```diff
@@ -224,302 +191,201 @@ def __init__(self, hdf_fpath):
 class ShellInfoWidget:
-    """The Shell Info Widget to explore abundances in different shells.
-
-    It consists of four interlinked table widgets - shells table; element count,
-    ion count and level count tables - allowing to explore fractional abundances
-    all the way from elements, to ions, to levels by clicking on the rows of
-    tables.
-    """
```
I aimed to make docstring more concise while maintaining clarity,.However, if the additional details are necessary for completeness, I’m open to refining the balance between brevity and informativeness.

andrewfullard commented on tardis/visualization/widgets/shell_info.py:
```diff
@@ -224,302 +191,201 @@ def __init__(self, hdf_fpath):
+        self.ion_count_table = pn.widgets.Tabulator(
+            pd.DataFrame(columns=["Species", "Frac. Ab."]),
+            layout="fit_data",
+            selectable=1,
```
Given that a lot of these calls to Tabulator have the same arguments, it may be reasonable to wrap this into a method with most of these arguments pre-set.

andrewfullard commented:
There are some typos in the markdown of the notebook example.

rudrakshnalbalwar commented:
hey @andrewfullard 
Thank you for the feedback, I've corrected the typos in markdown and refactored the repeated Tabulator calls into a separate method with pre-set arguments to improve code clarity and reduce redundancy

KasukabeDefenceForce commented on tardis/visualization/widgets/shell_info.py:
```diff
@@ -224,302 +191,199 @@ def __init__(self, hdf_fpath):
+                pn.Column(self.shells_title, self.shells_table, styles={'padding': '5px', 'border': '1px solid #ddd'}, width=320),
+                pn.Column(self.element_title, self.element_count_table, styles={'padding': '5px', 'border': '1px solid #ddd'}, width=240),
+                pn.Column(self.ion_title, self.ion_count_table, styles={'padding': '5px', 'border': '1px solid #ddd'}, width=280),
+                pn.Column(self.level_title, self.level_count_table, styles={'padding': '5px', 'border': '1px solid #ddd'}, width=280),
```
I would consider using a shared dictionary for column styles.

atharva-2001 commented on tardis/visualization/widgets/shell_info.py:
Can you add tests for this

rudrakshnalbalwar replied:
yes

atharva-2001 commented on .mailmap:
I think a merge conflict crept in, can you rebase this PR with the current master and resolve this conflict

atharva-2001 commented on tardis/visualization/widgets/shell_info.py:
```diff
@@ -225,280 +212,202 @@ def __init__(self, hdf_fpath):
+            self.update_ion_count_table(None)
+
+    def update_ion_count_table(self, event):
+        """Update ion table based on element selection"""
```
please look into the docstring format we use throughout tardis. You can look at the  tardis documentation for this

the tardis/base.py is a good example.

andrewfullard commented on tardis/visualization/widgets/tests/test_shell_info.py:
```diff
@@ -204,6 +361,16 @@ def test_selection_on_element_count_table(
+            The BaseShellInfo instance with expected data.
+        shell_info_widget : ShellInfoWidget
+            The widget to test.
+        """
```
We don't need docstrings for our tests

rudrakshnalbalwar replied:
sorry for that, I will remove them

atharva-2001 commented on tardis/visualization/widgets/tests/test_shell_info.py:
```diff
@@ -153,64 +164,144 @@ def shell_info_widget(self, base_shell_info, monkeysession):
+    def test_get_panel(self, shell_info_widget):
+ 
+        panel = shell_info_widget.get_panel()
+        assert panel is shell_info_widget.layout
```
panel is a library, so renaming variable this will be better

atharva-2001 commented on tardis/visualization/widgets/util.py:
```diff
@@ -4,9 +4,35 @@
+        elif isinstance(width, (int, float)):
+            return int(width)
+    except (AttributeError, ValueError):
+        pass
```
you should just not just catch errors and pass

rudrakshnalbalwar commented:
[shell_info](https://tardis-sn.github.io/tardis/pull/2990/analysing_tardis_outputs/visualization/how_to_generating_widgets.html#Shell-Info-Widget)
link to embedded widget

rudrakshnalbalwar commented:
Hey @andrewfullard, @atharva-2001, @KasukabeDefenceForce, @jamesgillanders, @wkerzendorf 
I've submitted the link for embedded widget (shell_info). Sorry for the delay! Let me know if you have any feedback on the implementation

atharva-2001 commented:
https://github.com/user-attachments/assets/bc4d7db4-a732-4b84-8df4-5e6478f430d0

the widget disappears occasionally unless I zoom in/out- is this an issue with panel? can you solve it?

rudrakshnalbalwar commented:
> Screen.Recording.2025-05-02.at.4.56.48.PM.mov 
> the widget disappears occasionally unless I zoom in/out- is this an issue with panel? can you solve it?

Hi @atharva-2001,
Thanks for the info on the disappearing widget. I've tested the widget on different browsers (Brave, Chrome, Firefox) while zooming in and out, but I haven't encountered the widget disappearing issue. Would you mind sharing which browser you're seeing this in? Knowing your specific browser might give me a clue about what I might be missing in my testing

andrewfullard commented:
Thank you very much for your contribution. We close GSOC-related pull requests after contributors are announced. You are welcome to contribute to TARDIS any time.
